Mathlib/Tactic/Linter/CommandStart.lean

```diff
@@ -0,0 +1,329 @@
+/-
+Copyright (c) 2025 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+
+import Mathlib.Init
+
+/-!
+#  The `commandStart` linter
+
+The `commandStart` linter emits a warning if
+* a command does not start at the beginning of a line;
+* the "hypotheses segment" of a declaration does not coincide with its pretty-printed version.
+-/
+
+open Lean Elab Command
+
+namespace Mathlib.Linter
+
+/--
+The `commandStart` linter emits a warning if
+* a command does not start at the beginning of a line;
+* the "hypotheses segment" of a declaration does not coincide with its pretty-printed version.
+
+In practice, this makes sure that the spacing in a typical declaration looks like
+```lean
+example (a : Nat) {R : Type} [Add R] : <not linted part>
+```
+as opposed to
+```lean
+example (a: Nat) {R:Type}  [Add  R] : <not linted part>
+```
+-/
+register_option linter.style.commandStart : Bool := {
+  defValue := false
+  descr := "enable the commandStart linter"
+}
+
+/-- If the `linter.style.commandStart.verbose` is `true`, the `commandStart` linter
+reports some helpful diagnostic information. -/
+register_option linter.style.commandStart.verbose : Bool := {
+  defValue := false
+  descr := "enable the commandStart linter"
+}
+
+/-- `lintUpTo stx` returns the position up until the `commandStart` linter checks the formatting.
+This is every declaration until the type-specification, if there is one, or the value,
+as well as all `variable` commands.
+-/
+def lintUpTo (stx : Syntax) : Option String.Pos :=
+  if let some cmd := stx.find? (·.isOfKind ``Parser.Command.declaration) then
+    if let some ind := cmd.find? (·.isOfKind ``Parser.Command.inductive) then
+      match ind.find? (·.isOfKind ``Parser.Command.optDeclSig) with
+      | none => dbg_trace "unreachable?"; none
+      | some sig => sig.getTailPos?
+    else
+    match cmd.find? (·.isOfKind ``Parser.Term.typeSpec) with
+      | some s => s.getPos?
+      | none => match cmd.find? (·.isOfKind ``Parser.Command.declValSimple) with
+        | some s => s.getPos?
+        | none => none
+  else if stx.isOfKind ``Parser.Command.variable then
+    stx.getTailPos?
+  else none
+
+/--
+A `FormatError` is the main structure for keeping track of how different the user syntax is
+from the pretty-printed version of itself.
+
+It contains information about position within an ambient string of where the exception lies.
+-/
+structure FormatError where
+  /-- The distance to the end of the source string, as number of characters. -/
+  srcNat : Nat
+  /-- The distance to the end of the source string, as number of string positions. -/
+  srcEndPos : String.Pos
```

Comment:
I don't think you want both of these.

Reply:
I find it surprisingly tricky to convert between the characters in the string and the string positions.

Moreover, the information on where the pretty-printed string and the user-provided string differs also involves a global shift by the initial position of the surface syntax.

So, while I agree that the information is redundant and could be teased off the `FileMap` and various other positions embedded in `Syntax`, I still found that maintaining a simple state of these numbers seemed easier.